schemes.go

```diff
@@ -1,70 +1,1151 @@
+	}
+	index := SchemeManagerIndex(make(map[string]SchemeFileHash))
+	if err = index.FromString(string(indexbts)); err != nil {
+		return nil, err, SchemeManagerStatusInvalidIndex
+	}
+	if err = conf.checkUnsignedFiles(dir, index); err != nil {
+		return nil, err, SchemeManagerStatusContentParsingError
+	}
+	return index, nil, SchemeManagerStatusValid
+}
+
+func (conf *Configuration) checkUnsignedFiles(dir string, index SchemeManagerIndex) error {
+	return common.WalkDir(dir, func(path string, info os.FileInfo) error {
+		relpath, err := filepath.Rel(dir, path)
+		if err != nil {
+			return err
+		}
+		schemepath := filepath.Join(index.Scheme(), relpath)
+		for _, ex := range sigExceptions {
+			if ex.MatchString(filepath.ToSlash(schemepath)) {
+				return nil
+			}
+		}
+
+		if info.IsDir() {
+			if !dirInScheme(index, schemepath) {
+				conf.Warnings = append(conf.Warnings, "Ignored dir: "+schemepath)
+			}
+		} else {
+			if _, ok := index[schemepath]; !ok {
+				conf.Warnings = append(conf.Warnings, "Ignored file: "+schemepath)
+			}
+		}
+
+		return nil
+	})
+}
+
+func downloadSignedFile(
+	transport *HTTPTransport, base, path string, hash SchemeFileHash,
+) ([]byte, error) {
+	b, err := transport.GetBytes(path)
+	if err != nil {
+		return nil, err
+	}
+	sha := sha256.Sum256(b)
+	if hash != nil && !bytes.Equal(hash, sha[:]) {
+		return nil, errors.Errorf("Signature over new file %s is not valid", path)
+	}
+	dest := filepath.Join(base, filepath.FromSlash(path))
+	if err = common.EnsureDirectoryExists(filepath.Dir(dest)); err != nil {
+		return nil, err
+	}
+	return b, common.SaveFile(dest, b)
+}
+
+func downloadFile(transport *HTTPTransport, base, path string) ([]byte, error) {
+	return downloadSignedFile(transport, base, path, nil)
+}
+
+func dirInScheme(index SchemeManagerIndex, dir string) bool {
+	for indexpath := range index {
+		if strings.HasPrefix(indexpath, dir) {
+			return true
+		}
+	}
+	return false
+}
+
+func downloadScheme(url string) (Scheme, error) {
+	if url[len(url)-1] == '/' {
+		url = url[:len(url)-1]
+	}
+
+	var filenames = common.SchemeFilenames
+	for _, filename := range common.SchemeFilenames {
+		if strings.HasSuffix(url, "/"+filename) {
+			filenames = []string{filename}
+		}
+	}
+
+	var scheme Scheme
+	for _, filename := range filenames {
+		u := url
+		if strings.HasSuffix(url, "/"+filename) {
+			u = url[:len(url)-1-len(filename)]
+		}
+		b, err := NewHTTPTransport(u, true).GetBytes(filename)
+		if err != nil {
+			if err.(*SessionError).RemoteStatus == 404 {
+				continue
+			}
+			return nil, err
+		}
+
+		_, typ, err := common.SchemeInfo(filename, b)
+		if err != nil {
+			return nil, err
+		}
+		scheme = newScheme(SchemeType(typ))
+		return scheme, common.Unmarshal(filename, b, scheme)
+	}
+
+	return nil, errors.New("no scheme description file found")
+}
+
+func tempSchemeCopy(scheme Scheme) (string, string, error) {
+	dir, err := ioutil.TempDir("", "tempscheme")
```

Comment:
The use of this function causes on Android the error `*os.PathError mkdir /data/local/tmp/tempscheme688734974: permission denied`. Probably Android does not allow apps to write to its default `/tmp` dir.

Reply:
Fixed